integration-tests/chaos/congestion/simulator.go

```diff
@@ -0,0 +1,484 @@
+				o.GasLimit = s.Input.GasLimit
+				o.GasTipCap = big.NewInt(0).SetUint64(s.expectedCurrentTip.Load())
+				o.GasFeeCap = big.NewInt(0).Add(big.NewInt(0).SetUint64(s.expectedCurrentBaseFee.Load()), o.GasTipCap)
+			})
+			if err != nil {
+				s.lggr.Errorw("failed to emit log to fill block", "error", err)
+			}
+		}
+	}
+}
+
+func (s *Simulator) reportObservationOnNewHead(ctx context.Context, header *types.Header) error {
+	congestion := float64(header.GasUsed) / float64(header.GasLimit)
+	baseFeePercent := float64(header.BaseFee.Uint64()) / float64(s.Input.InitialBaseFeePerGas) * 100
+	block, err := s.client.Client.BlockByNumber(ctx, header.Number)
+	if err != nil {
+		return fmt.Errorf("failed to fetch block by number: %w", err)
+	}
+	tipPercent := float64(s.avgTip(block).Uint64()) / float64(s.InitialTipPerGas) * 100
+
+	state := chainState{
+		PhaseName:           s.phase.Load().(phaseType),
+		Congestion:          congestion,
+		TipDeltaPercent:     tipPercent - 100,
+		BaseFeeDeltaPercent: baseFeePercent - 100,
+	}
+	if s.observationsHandler != nil {
+		s.observationsHandler(state)
+	}
+
+	s.lggr.Debugw("New chain state report", "chainState", state)
+	return nil
+}
+
+func (s *Simulator) avgTip(block *types.Block) *big.Int {
+	if block.Transactions().Len() == 0 {
+		return big.NewInt(0)
+	}
+	total := big.NewInt(0)
+	for _, tx := range block.Transactions() {
+		total = total.Add(tx.EffectiveGasTipValue(block.BaseFee()), total)
+	}
+
+	return big.NewInt(0).Div(total, big.NewInt(int64(block.Transactions().Len())))
+}
+
+func (s *Simulator) listenHeadsUntilCanceled(ctx context.Context, onNewBlock func(ctx context.Context, head *types.Header) error) {
+	for {
+		err := s.listenHeads(ctx, onNewBlock)
+		if err != nil {
+			if ctx.Err() != nil {
+				return
+			}
+
+			s.lggr.Errorw("failed to run onNewBlock", "error", err)
+		}
+		time.Sleep(backoffPeriodOnError)
+	}
+}
+
+func (s *Simulator) listenHeads(ctx context.Context, oneNewHead func(ctx context.Context, head *types.Header) error) error {
+	heads := make(chan *types.Header)
+	sub, err := s.client.Client.SubscribeNewHead(ctx, heads)
+	if err != nil {
+		return fmt.Errorf("failed to subscribe to new heads: %w", err)
+	}
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case head, ok := <-heads:
+			if !ok {
+				return errors.New("heads channel closed unexpectedly")
+			}
+			err = oneNewHead(ctx, head)
+			if err != nil {
+				if ctx.Err() != nil {
+					return ctx.Err()
+				}
+				s.lggr.Errorw("failed to run oneNewHead", "error", err)
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-sub.Err():
+			return fmt.Errorf("subscription error: %w", err)
+		}
+	}
+}
+
+func (s *Simulator) estimatePayloadSizeToFillBlock(ctx context.Context) (payloadSize int, err error) {
+	defer func() {
+		if rec := recover(); rec != nil {
+			err = fmt.Errorf("failed to estimate payload size to fill block: panic: %v", rec)
+		}
+	}()
```

Comment:
is it really necessary to use recover?

Reply:
I've used panics inside the search function for convenience, as the signature of that function does not allow us to easily abort it in case of an error.  
The `estimatePayloadSizeToFillBlock` signature allows to return an error, so it seems cleaner to me to recover the panic and return it as an error.